### ISSUE TITLE
Determine how the Beat and Bassline Editor should work

### ISSUE THREAD

**Issue by RebeccaDeField** [LMMS/lmms#2636]
From what people have been saying, it seems like the current BB editor is not ideal, and it's more from a technical standpoint than just aesthetic.

> What would the Beat and Bassline editor ideally look/work like? Maybe after we have that established, we can then come up with a solution based on those standards. ~ Rebecca

> We're still trying to determine what it should work like... I think ideally, the patterns would all grow and shrink together (something I'm trying to tackle right now). Then perhaps a way to prevent squishing the steps smaller than their natural size. I'll keep you posted as to my progress. This code is all over the place, so this is taking much more time than I had anticipated. ~ @tresf

**Comment by tresf** [comment]
Related: #2443

**Comment by tresf** [comment]
From a step-size perspective, I think all are in agreement that the steps need to be a predicable size, so that the pixmaps can be rendered correctly.

From a UI perspective there are some options... 
 * Make BB editor more like the song-editor in the fact that it can have variable length patterns and all patterns are positioned and looped based on their location in the BB timeline.
  * Currently the BB editor stretches patterns to the BB editor's width
  * Preventing patterns from stretching may mean we need a scroll area like the piano roll
  * Scroll area suggest we have a zoom factor as well
 * Keep BB editor fixed 
  * This is likely to be faster to implement from a code perspective
  * All patterns get adjusted to the BB editor's max pattern width
  * Steps never appear under a certain size
  * BB editor can quickly become wider than the average screen if long BB patterns are chosen

I had tried option 2 above but was having a hard time keeping the track widths in sync with each-other.  There seemed to be a missing way to quick expand/contract a pattern.

**Comment by liftarn** [comment]
On feature that would be nice to have is syncopated beats. Instead of just 1234 you would have 1a2a3a4a. You can do it as it is by either double the bpm and use half of them or going into the piano roll view, but neither of them is very user friendly.

**Comment by musikBear** [comment]
@liftarn you mean *sving* right?
I hope #2255 will be in consideration. Building continuous sample in succession would open complete new powerful ways of using lmms.

**Comment by liftarn** [comment]
@musikBear Sving is not something I'm familiar with. Swing I know what it is, but that is changing the timing of the play. I'm talking about adding beats at the positions between the 4/4 beats. So instead of 1 2 3 4 you have 1 and 2 and 3 and 4 and.

**Comment by RebeccaDeField** [comment]
Can we make the Beat and Bassline editor have a minimum width? It automatically opens in a size that squishes the graphics, but if we could have a minimum size that it scales to just as some of the other windows in LMMS do, I think it would greatly improve the interface.

![bbeditor](https://cloud.githubusercontent.com/assets/7960169/15685678/aff77fe4-2721-11e6-9046-8222e6041039.png)

**Comment by Umcaruje** [comment]
I'll look into this.

**Comment by Umcaruje** [comment]
@RebeccaDeField made #2808 to fix that.

**Comment by doyousketch2** [comment]
This might help [3325](https://github.com/LMMS/lmms/pull/3325)

Did something similar with the time signature division in a little beatbox app I made,
to show my kid how audio programs, such as LMMS, run.
scratch.mit.edu/projects/34189066

**Comment by jasp00** [comment]
>Make BB editor more like the song-editor

Yes, this is the idea. A BB track should be like a song: with its own tempo, time signature, master volume, etc. The BB editor is just the song editor with a starting zoom of 1600% (try it), plus the ability to add steps. BB tracks should have only the instruments they use; instruments can be shared among BB tracks, even the song.
